### ISSUE TITLE
Limit download size of the Pulumi CLI?

### ISSUE THREAD

**Issue by slavafomin** [pulumi/actions#787]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Hello! It looks like this GHA is downloading a [100+ MB archive](https://get.pulumi.com/releases/sdk/pulumi-v3.48.0-linux-x64.tar.gz) that includes binaries for all possible languages/platforms, which is a poor idea from the performance standpoint. It would be better to download only required binaries. This will help to reduce network traffic significantly, especially considering that this download happens on every GHA invocation.

Thanks.

### Affected area/feature

CLI download.

**Comment by RobbieMcKinstry** [comment]
I've got to eat my words. I think you're right, the performance improvement is worthwhile, and there's no reason not to implement it. 

#837 implements an optimization that skips downloads if a compatible CLI version is present.

Right now, when Pulumi is installed, the default language hosts must also be installed. The Pulumi CLI has a subcommand for installing language plugins individually (which was designed to support e.g. YAML) but this subcommand is broken for the language plugins that are considered default (Node, Dotnet, Python, and Go). So unfortunately we couldn't install _just_ Node even if we wanted to. There are a few Nix packages to make this work if you happen to have a runner with Nix installed.

I think in combination with #804 we'll squeeze as much performance from these optimizations as possible. If we cache plugins, plus skip redownloading the CLI and language hosts where possible, we should end up without any installation steps on a hosted runner with warm cache.

**Comment by Frassle** [comment]
> Right now, when Pulumi is installed, the default language hosts must also be installed. ...

N.B. I expect we'll change the core pulumi distribution to fix this at some point. One day a pulumi cli release will just have the cli in it, and dynamically fetch language plugins the same as provider plugins. Probably not worth trying to do all the effort here to work around that for now.

**Comment by RobbieMcKinstry** [comment]
Yes, I totally agree, that's the path forward in the long run. But it'll take a bit to get there unfortunately.
